File: mac/Sources/YTRec/Engine/BinaryLocator.swift
import Foundation

/// 找 yt-dlp / ffmpeg 執行檔：優先用 App 內建，其次開發目錄與 Homebrew
enum BinaryLocator {
    enum Tool: String {
        case ytdlp = "yt-dlp"
        case ffmpeg = "ffmpeg"
    }

    static func url(for tool: Tool) -> URL? {
        var candidates: [URL] = []
        if let res = Bundle.main.resourceURL {
            candidates.append(res.appendingPathComponent("bin/\(tool.rawValue)"))
        }
        if let dev = ProcessInfo.processInfo.environment["LCF_BIN_DIR"] {
            candidates.append(URL(fileURLWithPath: dev).appendingPathComponent(tool.rawValue))
        }
        // swift run 開發情境：從執行檔往上找 Vendor/bin
        let exe = URL(fileURLWithPath: CommandLine.arguments[0]).resolvingSymlinksInPath()
        var dir = exe.deletingLastPathComponent()
        for _ in 0..<6 {
            candidates.append(dir.appendingPathComponent("Vendor/bin/\(tool.rawValue)"))
            dir.deleteLastPathComponent()
        }
        candidates.append(URL(fileURLWithPath: "/opt/homebrew/bin/\(tool.rawValue)"))
        candidates.append(URL(fileURLWithPath: "/usr/local/bin/\(tool.rawValue)"))

        return firstExecutable(candidates, isExecutable: FileManager.default.isExecutableFile(atPath:))
    }

    /// 從候選清單挑「第一個可執行」（純函式，可注入 isExecutable 以擺脫真檔系統）。
    /// 候選順序＝內嵌 bin → 開發環境變數 → Vendor/bin → Homebrew，故「內嵌優先、找不到回 nil」由此保證。
    static func firstExecutable(_ candidates: [URL], isExecutable: (String) -> Bool) -> URL? {
        candidates.first { isExecutable($0.path) }
    }

    static var missingTools: [String] {
        var missing: [String] = []
        if url(for: .ytdlp) == nil { missing.append("yt-dlp") }
        if url(for: .ffmpeg) == nil { missing.append("ffmpeg") }
        return missing
    }
}

File: windows/YtRec.Core/BinaryLocator.cs
namespace YtRec.Core;

/// <summary>Locates the yt-dlp / ffmpeg executables. Mirrors mac BinaryLocator.swift.
/// Search order: bundled (app dir/vendor/bin) → YTREC_BIN_DIR env → PATH. Cross-platform
/// (adds .exe on Windows). The pure helpers are unit-tested without touching the real FS.</summary>
public static class BinaryLocator
{
    public enum Tool { YtDlp, Ffmpeg }

    /// <summary>Platform executable filename for a tool (yt-dlp / yt-dlp.exe).</summary>
    public static string ExecutableName(Tool tool)
    {
        var stem = tool == Tool.YtDlp ? "yt-dlp" : "ffmpeg";
        return OperatingSystem.IsWindows() ? stem + ".exe" : stem;
    }

    /// <summary>Ordered candidate paths for a tool. baseDir defaults to the app's base directory.</summary>
    public static IReadOnlyList<string> Candidates(Tool tool, string? baseDir = null,
        string? binDirEnv = null, string? pathEnv = null)
    {
        baseDir ??= AppContext.BaseDirectory;
        binDirEnv ??= Environment.GetEnvironmentVariable("YTREC_BIN_DIR");
        pathEnv ??= Environment.GetEnvironmentVariable("PATH");
        var name = ExecutableName(tool);

        var list = new List<string>
        {
            Path.Combine(baseDir, "vendor", "bin", name), // bundled next to the app
            Path.Combine(baseDir, name),                  // alongside the exe
        };
        if (!string.IsNullOrEmpty(binDirEnv))
            list.Add(Path.Combine(binDirEnv, name));       // dev/override
        if (!string.IsNullOrEmpty(pathEnv))
            foreach (var dir in pathEnv.Split(Path.PathSeparator, StringSplitOptions.RemoveEmptyEntries))
                list.Add(Path.Combine(dir, name));         // system PATH
        return list;
    }

    /// <summary>First candidate the predicate accepts (pure; inject the FS check for tests).</summary>
    public static string? FirstExecutable(IEnumerable<string> candidates, Func<string, bool> exists)
        => candidates.FirstOrDefault(exists);

    /// <summary>Resolve a tool's path against the real filesystem, or null if not found.</summary>
    public static string? Resolve(Tool tool, string? baseDir = null)
        => FirstExecutable(Candidates(tool, baseDir), File.Exists);

    /// <summary>Names of tools that could not be located (stable order: yt-dlp, ffmpeg).</summary>
    public static IReadOnlyList<string> MissingTools(string? baseDir = null)
    {
        var missing = new List<string>();
        if (Resolve(Tool.YtDlp, baseDir) is null) missing.Add(ExecutableName(Tool.YtDlp));
        if (Resolve(Tool.Ffmpeg, baseDir) is null) missing.Add(ExecutableName(Tool.Ffmpeg));
        return missing;
    }
}
